File: adt2unit/pas/update_extra_fine_effects.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure update_extra_fine_effects;

var
  chan,eLo,eHi,
  eLo2,eHi2: Byte;

begin
  For chan := 1 to songdata.nm_tracks do
    begin
      eLo  := LO(effect_table[chan]);
      eHi  := HI(effect_table[chan]);
      eLo2 := LO(effect_table2[chan]);
      eHi2 := HI(effect_table2[chan]);

      Case eLo of
        ef_extended2+ef_fix2+ef_ex2_GlVolSldUpXF:
          global_volume_slide(eHi,BYTE_NULL);

        ef_extended2+ef_fix2+ef_ex2_GlVolSldDnXF:
          global_volume_slide(BYTE_NULL,eHi);

        ef_extended2+ef_fix2+ef_ex2_VolSlideUpXF:
          volume_slide(chan,eHi,0);

        ef_extended2+ef_fix2+ef_ex2_VolSlideDnXF:
          volume_slide(chan,0,eHi);

        ef_extended2+ef_fix2+ef_ex2_FreqSlideUpXF:
          portamento_up(chan,eHi,nFreq(12*8+1));

        ef_extended2+ef_fix2+ef_ex2_FreqSlideDnXF:
          portamento_down(chan,eHi,nFreq(0));

        ef_ExtraFineArpeggio:
          arpeggio(chan);

        ef_ExtraFineVibrato:
          If NOT vibr_table[chan].fine then
            vibrato(chan);

        ef_ExtraFineTremolo:
          If NOT trem_table[chan].fine then
            tremolo(chan);
      end;

      Case eLo2 of
        ef_extended2+ef_fix2+ef_ex2_GlVolSldUpXF:
          global_volume_slide(eHi2,BYTE_NULL);

        ef_extended2+ef_fix2+ef_ex2_GlVolSldDnXF:
          global_volume_slide(BYTE_NULL,eHi2);

        ef_extended2+ef_fix2+ef_ex2_VolSlideUpXF:
          volume_slide(chan,eHi2,0);

        ef_extended2+ef_fix2+ef_ex2_VolSlideDnXF:
          volume_slide(chan,0,eHi2);

        ef_extended2+ef_fix2+ef_ex2_FreqSlideUpXF:
          portamento_up(chan,eHi2,nFreq(12*8+1));

        ef_extended2+ef_fix2+ef_ex2_FreqSlideDnXF:
          portamento_down(chan,eHi2,nFreq(0));

        ef_ExtraFineArpeggio:
          arpeggio2(chan);

        ef_ExtraFineVibrato:
          If NOT vibr_table2[chan].fine then
            vibrato2(chan);

        ef_ExtraFineTremolo:
          If NOT trem_table2[chan].fine then
            tremolo2(chan);
      end;
    end;
end;

File: adt2unit/update_extra_fine_effects.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

/*static*/ void update_extra_fine_effects (void)
{
  for (uint8_t chan = 0; chan < songdata.nm_tracks; chan++)
  {
    uint8_t eLo  = effect_table[chan] & 0xFF;
    uint8_t eHi  = effect_table[chan] >> 8;
    uint8_t eLo2 = effect_table2[chan] & 0xFF;
    uint8_t eHi2 = effect_table2[chan] >> 8;

    switch (eLo)
    {
      case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSldUpXF:
        global_volume_slide (eHi, BYTE_NULL);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSldDnXF:
        global_volume_slide (BYTE_NULL, eHi);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_VolSlideUpXF:
        volume_slide (chan + 1, eHi, 0);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_VolSlideDnXF:
        volume_slide (chan + 1, 0, eHi);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_FreqSlideUpXF:
        portamento_up (chan + 1, eHi, nFreq (12*8+1));
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_FreqSlideDnXF:
        portamento_down (chan + 1, eHi, nFreq (0));
        break;

      case ef_ExtraFineArpeggio:
        arpeggio (chan + 1);
        break;

      case ef_ExtraFineVibrato:
        if (!vibr_table[chan].fine)
          vibrato (chan + 1);
        break;

      case ef_ExtraFineTremolo:
        if (!trem_table[chan].fine)
          tremolo (chan + 1);
        break;

      default:
        break;
    }

    switch (eLo2)
    {
      case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSldUpXF:
        global_volume_slide (eHi2, BYTE_NULL);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSldDnXF:
        global_volume_slide (BYTE_NULL, eHi2);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_VolSlideUpXF:
        volume_slide (chan + 1, eHi2, 0);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_VolSlideDnXF:
        volume_slide (chan + 1, 0, eHi2);
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_FreqSlideUpXF:
        portamento_up (chan + 1, eHi2, nFreq (12*8+1));
        break;

      case ef_Extended2 + ef_fix2 + ef_ex2_FreqSlideDnXF:
        portamento_down (chan + 1, eHi2, nFreq (0));
        break;

      case ef_ExtraFineArpeggio:
        arpeggio2 (chan + 1);
        break;

      case ef_ExtraFineVibrato:
        if (!vibr_table2[chan].fine)
          vibrato2 (chan + 1);
        break;

      case ef_ExtraFineTremolo:
        if (!trem_table2[chan].fine)
          tremolo2 (chan + 1);
        break;

      default:
        break;
    }
  }
}
